lib/Transforms/GenStateStruct.cpp

```diff
@@ -0,0 +1,1021 @@
+    delayedTy.push_back(rstruct.regType);
+  }
+
+  // Generate the pointer type for the submodule part.
+  for (auto inst : llvm::make_early_inc_range(op.getOps<hw::InstanceOp>())) {
+    if (auto refedModule =
+            dyn_cast<HWModuleOp>(*instGraph.getReferencedModule(inst)))
+      innerStructTy = LLVM::LLVMStructType::getIdentified(
+          ctx, addPrefixToName(refedModule.getName()));
+    else
+      innerStructTy = getExternModulePortTy(
+          ctx, false,
+          dyn_cast<HWModuleExternOp>(*instGraph.getReferencedModule(inst)));
+    packedStructTy.push_back(LLVM::LLVMPointerType::get(innerStructTy));
+  }
+
+  // Combine all container to create the final struct type.
+  clonedPortTy.insert(clonedPortTy.end(), delayedTy.begin(), delayedTy.end());
+  clonedPortTy.insert(clonedPortTy.end(), packedStructTy.begin(),
+                      packedStructTy.end());
+
+  // The NewIdentified struct type has an explicit name and type. If we want to
+  // change the struct type to add or remove some members,we need to remove the
+  // whole struct definition beacuse the Indentified Struct does not support
+  // modify structure directly.
+  return LLVM::LLVMStructType::getNewIdentified(
+      ctx, addPrefixToName(op.getName()), clonedPortTy,
+      /*isPacked=*/true);
+}
+
+LLVM::GlobalOp GenStateStruct::createGlobalStruct(Location loc,
+                                                  OpBuilder &builder,
+                                                  ModuleOp module,
+                                                  StringRef name,
+                                                  LLVM::LLVMStructType type) {
+  LLVM::GlobalOp global;
+
+  // Create an empty global struct which is uninitialized.
+  if (!(global = module.lookupSymbol<LLVM::GlobalOp>(name))) {
+    OpBuilder::InsertionGuard insertGuard(builder);
+    builder.setInsertionPointToStart(module.getBody());
+    global = builder.create<LLVM::GlobalOp>(loc, type, /*isConstant=*/false,
+                                            LLVM::Linkage::Internal, name,
+                                            Attribute(),
+                                            /*alignment=*/0, /*addrspace=*/0);
+  }
+  return global;
+}
+
+// TODO: initialize the global struct with completely random value. Maybe it is
+// a new option for this pass.
+void GenStateStruct::initializeGlobalStruct(
+    OpBuilder &builder, HWModuleOp op, LLVM::GlobalOp globalOp,
+    LLVM::LLVMStructType type, InstanceGraph &instGraph,
+    SmallVector<PortInfoCollect> &pInfo, SmallVector<RegInfoCollect> &rInfo) {
+  // Initialize the global struct with specified information.
+  Location loc = globalOp.getLoc();
+  MLIRContext *ctx = builder.getContext();
+  Region &region = globalOp.getInitializerRegion();
+  Block *block = builder.createBlock(&region);
+  builder.setInsertionPoint(block, block->begin());
+
+  LLVM::LLVMStructType innerStructTy;
+  IntegerType i1Ty = IntegerType::get(ctx, 1);
```

Comment:
I think builder has a convenience function for IntegerType.

Reply:
Yes, this should be changed